src/message_bus_fuzz.zig

```diff
@@ -0,0 +1,914 @@
+                    return .done;
+                }
+            },
+            .send => |operation| {
+                const sender = io.connections.getPtr(operation.socket).?;
+                assert(!sender.closed);
+                assert(sender.pending_send);
+                sender.pending_send = false;
+
+                if (io.prng.chance(io.options.send_error_probability)) {
+                    const result: SendError!usize = io.prng.error_uniform(SendError);
+                    completion.callback(completion.context, completion, &result);
+
+                    // If there is a pending recv(), we need to make sure that it fails (rather than
+                    // potentially stalling and preventing MessageBus from calling shutdown/close).
+                    sender.closed = true;
+                    return .done;
+                }
+
+                if (sender.shutdown_send) {
+                    const result: SendError!usize = error.BrokenPipe;
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
+
+                const send_size = if (io.prng.chance(io.options.send_partial_probability))
+                    io.prng.range_inclusive(usize, 0, operation.buffer.len)
+                else
+                    operation.buffer.len;
+
+                const send_buffer = operation.buffer[0..send_size];
+                try sender.sending.appendSlice(gpa, send_buffer);
+
+                if (io.prng.chance(io.options.send_corrupt_probability)) {
+                    if (send_buffer.len == 0) {
+                        try sender.sending.append(gpa, io.prng.int(u8));
+                    } else {
+                        const corrupt_byte = io.prng.index(send_buffer);
+                        sender.sending.items[sender.sending.items.len - corrupt_byte - 1] ^=
+                            io.prng.bit(u8);
+                    }
+                }
+
+                const result: SendError!usize = send_size;
+                completion.callback(completion.context, completion, &result);
+            },
+            .recv => |operation| {
+                const receiver = io.connections.getPtr(operation.socket).?;
+                assert(receiver.pending_recv);
+                // If we had a send() fail, then we must avoid requeueing this operation.
+                maybe(receiver.closed);
+
+                if (receiver.closed or io.prng.chance(io.options.recv_error_probability)) {
+                    const result: RecvError!usize = io.prng.error_uniform(RecvError);
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
+
+                if (receiver.shutdown_recv) {
+                    const result: RecvError!usize = 0;
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
```

Comment:
Similar here, I'd expect that it's possible that recv initiated before shutdown can complete after. Actually, I'd expect that to be the normal case? Shutdown flushes everything already submitted, and then closes, rather than abruptly cutting

Reply:
Ooh, good catch! I was going off of the man page which says:

> If `how` is `SHUT_RD`, further receptions will be  disallowed.

...but I think that is misleading, and you can still receive anything that had already been buffered!

Reply:
I think another guiding principle is that we shouldn't cut it too close to the manpage, if we can impose _less_ restrictions on semantics, it might better to do that!